### PR TITLE
AS: Balancer V2 batchSwap

### PR DESCRIPTION
**Pull request by dorothy-zbornak**
## Description

Based on #445 (ty @johngrantuk), forked for integration testing and tuning.

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

**Comment by dorothy-zbornak** [comment]
@johngrantuk can't seem to assign you to the PR so tagging you here.

File: packages/asset-swapper/src/utils/market_operation_utils/pools_cache/sgPoolDataService.ts [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        }
+    }
+`;
+
+const QUERY_BY_CHAIN_ID: { [chainId: number]: string } = {
+    [ChainId.Mainnet]: queryWithLinear,
+    [ChainId.Polygon]: queryWithOutLinear,
+};
+
+const DEFAULT_MAX_POOLS_FETCHED = 96;
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/src/utils/market_operation_utils/pools_cache/sgPoolDataService.ts:79]: This number really impacts how long AS takes to boot up so don't set it too high.

File: contracts/zero-ex/contracts/src/transformers/bridges/BridgeProtocols.sol [hunk 2]
```diff
@@ -27,29 +27,30 @@ library BridgeProtocols {
...
+    uint128 internal constant MAKERPSM        = 16;
+    uint128 internal constant BALANCERV2      = 17;
+    uint128 internal constant UNISWAPV3       = 18;
+    uint128 internal constant KYBERDMM        = 19;
+    uint128 internal constant CURVEV2         = 20;
+    uint128 internal constant LIDO            = 21;
+    uint128 internal constant CLIPPER         = 22; // Not used: Clipper is now using PLP interface
+    uint128 internal constant AAVEV2          = 23;
+    uint128 internal constant COMPOUND        = 24;
+    uint128 internal constant BALANCERV2BATCH = 25;
```
Comment (dorothy-zbornak) [inline contracts/zero-ex/contracts/src/transformers/bridges/BridgeProtocols.sol:55]: New Protocol ID. `BALANCERV2` is still used by BeethovenX so couldn't upgrade that in place.

File: packages/asset-swapper/src/utils/market_operation_utils/pools_cache/balancer_v2_utils_new.ts [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+     * @returns Swap data for pair consisting of assets and swap steps for ExactIn and ExactOut swap types.
+     */
+    private _getPoolPairSwapInfo(pools: PoolDictionary, takerToken: string, makerToken: string): BalancerSwaps {
+        /*
+        Uses Balancer SDK to construct available paths for pair.
+        Paths can be direct, i.e. both tokens are in same pool or multihop.
+        Will also create paths for the new Balancer Linear pools.
+        These are returned in order of available liquidity which is useful for filtering.
+        */
+        const maxPools = 4;
```
Comment (mzhu25) [inline packages/asset-swapper/src/utils/market_operation_utils/pools_cache/balancer_v2_utils_new.ts:134]: private static member instead?

File: packages/asset-swapper/src/utils/market_operation_utils/pools_cache/balancer_v2_utils_new.ts [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+            pools,
+            maxPools,
+        );
+
+        if (paths.length === 0) {
+            return EMPTY_BALANCER_SWAPS;
+        }
+
+        // Convert paths data to swap information suitable for queryBatchSwap. Only use top 3 liquid paths
+        return formatSwaps(paths.slice(0, 3));
```
Comment (mzhu25) [inline packages/asset-swapper/src/utils/market_operation_utils/pools_cache/balancer_v2_utils_new.ts:148]: pull maxPaths out into a constant too

File: packages/asset-swapper/src/utils/market_operation_utils/pools_cache/pair_swaps_cache.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        const timeout = new Promise<BalancerSwaps>(resolve => setTimeout(resolve, timeoutMs, []));
+        return Promise.race([this._getAndSaveFreshSwapInfoForPairAsync(takerToken, makerToken), timeout]);
+    }
+
+    public getCachedSwapInfoForPair(
+        takerToken: string,
+        makerToken: string,
+        ignoreExpired: boolean = true,
+    ): BalancerSwaps | undefined {
+        const key = JSON.stringify([takerToken, makerToken]);
```
Comment (mzhu25) [inline packages/asset-swapper/src/utils/market_operation_utils/pools_cache/pair_swaps_cache.ts:44]: should we be lowercasing the token addresses in this class? I don't remember whether we do it upstream somewhere

File: packages/asset-swapper/contracts/src/ERC20BridgeSampler.sol [hunk 6]
```diff
@@ -45,30 +44,7 @@ import "./UniswapV3Sampler.sol";
...
-    MooniswapSampler,
-    MultiBridgeSampler,
-    NativeOrderSampler,
-    ShellSampler,
-    SmoothySampler,
-    TwoHopSampler,
-    UniswapSampler,
-    UniswapV2Sampler,
-    UniswapV3Sampler,
-    UtilitySampler
```
[the remark this answers is not in the document]
Comment (dorothy-zbornak) [reply packages/asset-swapper/contracts/src/ERC20BridgeSampler.sol]: sorry, me messing with CI to get it to compile *anything*

File: packages/asset-swapper/src/utils/market_operation_utils/pools_cache/pair_swaps_cache.ts (hunk 5 above)
Comment (dorothy-zbornak) [reply packages/asset-swapper/src/utils/market_operation_utils/pools_cache/pair_swaps_cache.ts:44]: I think tokens come in lowercased and, so long as we don't mess up token constants, intermediate tokens should be too.

**Review by dorothy-zbornak — COMMENTED** [review]
wow great job, don't ever quit

**Pull request merged**